docs/proposals/controller-ref.md

```diff
@@ -0,0 +1,131 @@
+Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
+
+# Non goals
+
+Cascading deletion is not a goal of this effort. Cascading deletion will use `ownerReferences`, which is a [separate effort](garbage-collection.md).
+
+`ControllerRef` will extend `OwnerReference` and reuse machinery written for it (GarbageCollector, adoption/orphaning logic).
+
+# API and semantic changes
+
+There will be a new API field in the `OwnerReference` in which we will store an information if given owner is a managing controller:
+
+```
+OwnerReference {
+    …
+    Controller bool
+    …
+}
+```
+
+From now on by `ControllerRef` we mean an `OwnerReference` with `Controller=true`.
+
+Most controllers (all that manage collections of things defined by label selector) will have slightly changed semantics: currently controller owns an object if its selector matches object’s labels and if it doesn't notice an older controller of the same kind that also matches the object's labels, but after introduction of `ControllerReference` a controller will own an object iff selector matches labels and the `OwnerReference` with `Controller=true`points to it.
+
+If the owner's selector or owned object's labels change, the owning controller will be responsible for orphaning (clearing `Controller` field in the `OwnerReference` and/or deleting `OwnerReference` altogether) objects, after which adoption procedure (setting `Controller` field in one of `OwnerReferencec` and/or adding new `OwnerReferences`) might occur, if another controller has a selector matching.
+
+For debugging purposes we want to add an `adoptionTime` annotation prefixed with `kubernetes.io/` which will keep the time of last controller ownership transfer.
+
+# Upgrade/downgrade procedure
+
+Because `ControllerRef` will be a part of `OwnerReference` effort it will have the same upgrade/downgrade procedures.
+
+# Orphaning/adoption
+
+Because `ControllerRef` will be a part of `OwnerReference` effort it will have the same orphaning/adoption procedures.
+
+Controllers will orphan objects they own in two cases:
+* Change of label/selector causing selector to stop matching labels (executed by the controller)
+* Deletion of a controller with `Orphaning=true` (executed by the GarbageCollector)
+
+We will need a secondary orphaning mechanism in case of unclean controller deletion:
+* GarbageCollector will remove `ControllerRef` from objects that no longer points to existing controllers
+
+Controller will adopt (set `Controller` field in the `OwnerReference` that points to it) an object whose labels match its selector iff:
+* there are no `OwnerReferences` with `Controller` set to true in `OwnerReferences` array
+* `DeletionTimestamp` is not set
+and
+* Controller is the oldest controller of a given kind from ones which has a selector that matches labels of adopted object that does not have `DeletionTimestamp` set.
```

Comment:
I'm with @caesarxuchao here. How about using the pattern @erictune introduced in Jobs, where upon creation we add an additional selector which guarantees pod selection uniqueness? See [this code](https://github.com/kubernetes/kubernetes/blob/524c5b53614c8d1dfb6bf607f8674b71bb655180/pkg/registry/job/strategy.go#L74-L114).